sdk/core/core-rest-pipeline/documentation/pipeline.md

```diff
@@ -0,0 +1,220 @@
+# Policies
+
+A pipeline policy manipulates a request as it travels through the pipeline. It is conceptually a middleware that is allowed to modify the request before it is made as well as the response when it is received.
+
+```typescript
+export type SendRequest = (request: PipelineRequest) => Promise<PipelineResponse>;
+
+export interface PipelinePolicy {
+  name: string;
+  sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse>;
+}
+```
+
+It is given a string name so that consumers can express dependencies relative to existing policies. The `sendRequest` method uses a middleware pattern, which is familiar to JS developers.
+
+# Pipelines
+
+A `Pipeline` composes several policies and provides the ability to make a request that will flow through the policies in a specific order determined by their requirements. Which allows manipulating each request before and after it is made to the server.
+
+```typescript
+export interface AddPipelineOptions {
+  afterPhase?: PipelinePhase;
+  afterPolicies?: string[];
+  beforePolicies?: string[];
+  phase?: PipelinePhase;
+}
+
+export interface Pipeline {
+  addPolicy(policy: PipelinePolicy, options?: AddPipelineOptions): void;
+  clone(): Pipeline;
+  getOrderedPolicies(): PipelinePolicy[];
+  removePolicy(options: { name?: string; phase?: PipelinePhase }): PipelinePolicy[];
+  sendRequest(httpClient: HttpClient, request: PipelineRequest): Promise<PipelineResponse>;
+}
+```
+
+Dependencies expressed when calling `addPolicy` are not sensitive to order. The following code snippets have identical effects:
+
+```typescript
+pipeline.addPolicy(barPolicy);
+pipeline.addPolicy(fooPolicy, { afterPolicy: "barPolicy" });
+
+// The below is the same as the above.
+// Policies don't have to be part of the pipeline in order to be
+// referenced by before/afterPolicy.
+pipeline.addPolicy(fooPolicy, { afterPolicy: "barPolicy" });
+pipeline.addPolicy(barPolicy);
+```
+
+Phases are predefined buckets of activity that pre-defined policies inside core will fall into. They are currently a fixed set and not customizable by pipeline consumers.
+
+```typescript
+export type PipelinePhase = "Deserialize" | "Serialize" | "Retry" | "Sign";
+```
+
```

Comment:
I think a description of what kind of work gets done in each and what order they go in would be good